src/proof/lrat/lrat_proof.cpp

```diff
@@ -0,0 +1,199 @@
+  for (auto& c : usedClauses)
+  {
+    for (auto l : *(c.second))
+    {
+      if (l.getSatVariable() + 1 > maxVar)
+      {
+        maxVar = l.getSatVariable() + 1;
+      }
+    }
+  }
+  formStream << "p cnf " << maxVar << " " << usedClauses.size() << '\n';
+  for (auto ci : clauseOrder)
+  {
+    auto iterator = usedClauses.find(ci);
+    Assert(iterator != usedClauses.end());
+    for (auto l : *(iterator->second))
+    {
+      if (l.isNegated())
+      {
+        formStream << '-';
+      }
+      formStream << l.getSatVariable() + 1 << " ";
+    }
+    formStream << "0\n";
+  }
+  formStream.close();
+
+  std::ofstream dratStream(dratFilename);
+  dratStream << dratBinary;
+  dratStream.close();
+
+  // TODO(aozdemir) Add invocation of DRAT trim, once I get CMake to bundle it
+  // into CVC4 correctly.
+  Unimplemented();
+
+  std::ifstream lratStream(lratFilename);
+  LRATProof lratProof(lratStream);
+  remove(formulaFilename);
+  remove(dratFilename);
+  remove(lratFilename);
+  *this = std::move(lratProof);
+}
+
+std::istream& operator>>(std::istream& in, prop::SatLiteral& l)
+{
+  long int i;
+  in >> i;
+  l = prop::SatLiteral(labs(i), i < 0);
+  return in;
+}
+
+// This parser is implemented to parse the textual RAT format found in
+// "Efficient Certified RAT Verification", by Cruz-Filipe et. All
+LRATProof::LRATProof(std::istream& textualProof)
+{
+  for (size_t line = 0;; ++line)
+  {
+    LRATInstruction instr;
+
+    // Read beginning of instruction. EOF indicates that we're done.
+    size_t clauseIdx;
+    textualProof >> clauseIdx;
+    if (textualProof.eof())
+    {
+      return;
+    }
+
+    // Read the first word of the instruction. A 'd' indicates deletion.
+    std::string first;
+    textualProof >> first;
+    Trace("pf::lrat") << "First word: " << first << std::endl;
+    Assert(textualProof.good());
+    if (first == "d")
+    {
+      instr.kind = lratDeletion;
+      instr.deletionData.idxOfClause = clauseIdx;
+      while (true)
+      {
+        ClauseIdx di;
+        textualProof >> di;
+        Assert(textualProof.good());
+        if (di == 0)
+        {
+          break;
+        }
+        instr.deletionData.clauses.push_back(di);
+      }
+      std::sort(instr.deletionData.clauses.begin(),
+                instr.deletionData.clauses.end());
+    }
+    else
+    {
+      instr.kind = lratAddition;
+      instr.additionData.idxOfClause = clauseIdx;
+
+      // We need to reparse the first word as a literal to read the clause
+      // we're parsing. It ends with a 0;
+      std::istringstream firstS(first);
+      SatLiteral lit;
+      firstS >> lit;
+      Trace("pf::lrat") << "First lit: " << lit << std::endl;
+      Assert(!firstS.fail(), "Couldn't parse first literal from addition line");
+      for (; lit != 0; textualProof >> lit)
+      {
+        Assert(textualProof.good());
+        SatLiteral l(lit.getSatVariable() - 1, lit.isNegated());
+        instr.additionData.clause.push_back(l);
```

Comment:
[Optional] Suggestion: you can combine the above two lines to `instr.additionData.clause.emplace_back(lit.getSatVariable() - 1, lit.isNegated());`, avoiding a copy of the `SatLiteral`.

Reply:
Sounds good.